### ISSUE TITLE
Timeout when acquiring next swap chain texture

### ISSUE THREAD

**Issue by jpopesculian** [nannou-org/nannou#651]
Hey! I love nannou, but last week suddenly nannou seems to crash whenever I try to run a program. I'm running

* OS: Ubuntu 20.04.1 LTS
* Processor: Intel® Core™ i7-10510U CPU @ 1.80GHz × 8
* Graphics: Mesa Intel® UHD Graphics (CML GT2)
* Kernel: 5.4.0
* Windowing: X11

I should have the most up to date mesa drivers. I found this issue https://github.com/gfx-rs/wgpu-rs/issues/164 but the examples for `winit`, `wgpu` and `ash` all run fine. I don't know how to investigate the issue any further, so any help would greatly be appreciated.

Here's the error when I do `RUST_BACKTRACE=full cargo run --example draw`:

```
WARNING: vallium/llvmpipe is not a conformant vulkan implementation, testing use only.
INTEL-MESA: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0

thread 'main' panicked at 'failed to acquire next swapchain texture: TimeOut', /home/julian/Development/nannou/nannou/src/app.rs:1069:45
stack backtrace:
   0:     0x55cd9c16ea85 - backtrace::backtrace::libunwind::trace::h396c07d2071b43af
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x55cd9c16ea85 - backtrace::backtrace::trace_unsynchronized::h7aa0e4bb23d9c158
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x55cd9c16ea85 - std::sys_common::backtrace::_print_fmt::hd15ac5d4adcd355b
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x55cd9c16ea85 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hec5354be8ccc3ecc
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x55cd9c19601c - core::fmt::write::h3d34909eeb4f225b
                               at src/libcore/fmt/mod.rs:1076
   5:     0x55cd9c16b623 - std::io::Write::write_fmt::h1da287b3de55ed16
                               at src/libstd/io/mod.rs:1537
   6:     0x55cd9c171210 - std::sys_common::backtrace::_print::h4d206838e1ace354
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x55cd9c171210 - std::sys_common::backtrace::print::h1f778e9940ee5977
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x55cd9c171210 - std::panicking::default_hook::{{closure}}::h704403a56cbf5783
                               at src/libstd/panicking.rs:198
   9:     0x55cd9c170f5c - std::panicking::default_hook::ha4567a10dec4ef8d
                               at src/libstd/panicking.rs:218
  10:     0x55cd9c171847 - std::panicking::rust_panic_with_hook::h88a1f16ec8a7bb20
                               at src/libstd/panicking.rs:486
  11:     0x55cd9c17144b - rust_begin_unwind
                               at src/libstd/panicking.rs:388
  12:     0x55cd9c193ca1 - core::panicking::panic_fmt::hbddb7fe6f399b81a
                               at src/libcore/panicking.rs:101
  13:     0x55cd9c193ac3 - core::option::expect_none_failed::h60849c4323f09783
                               at src/libcore/option.rs:1272
  14:     0x55cd9b7a9acb - core::result::Result<T,E>::expect::hb280fcb90f7961f0
                               at /home/julian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/result.rs:963
  15:     0x55cd9b7e59e7 - nannou::app::run_loop::{{closure}}::ha2494c1b8b74c05d
                               at /home/julian/Development/nannou/nannou/src/app.rs:1069
  16:     0x55cd9b7ba803 - winit::platform_impl::platform::sticky_exit_callback::h7cf75c9307d6f82d
                               at /home/julian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/mod.rs:698
  17:     0x55cd9b7a74bc - winit::platform_impl::platform::x11::EventLoop<T>::run_return::h8947e3163665804b
                               at /home/julian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/x11/mod.rs:312
  18:     0x55cd9b7a84db - winit::platform_impl::platform::x11::EventLoop<T>::run::h6fcc1ddb5c0d5458
                               at /home/julian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/x11/mod.rs:390
  19:     0x55cd9b7ba3b8 - winit::platform_impl::platform::EventLoop<T>::run::h559ef7107084368b
                               at /home/julian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/mod.rs:645
  20:     0x55cd9b796c74 - winit::event_loop::EventLoop<T>::run::hf1b02f445395e661
                               at /home/julian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/event_loop.rs:149
  21:     0x55cd9b7e5410 - nannou::app::run_loop::h875b04b71c995c92
                               at /home/julian/Development/nannou/nannou/src/app.rs:1020
  22:     0x55cd9b7de1fe - nannou::app::Builder<M,E>::run::h500231b6cf7aa4ce
                               at /home/julian/Development/nannou/nannou/src/app.rs:471
  23:     0x55cd9b7de3c2 - nannou::app::SketchBuilder<E>::run::hb3560227d4858995
                               at /home/julian/Development/nannou/nannou/src/app.rs:496
  24:     0x55cd9b7ac0c0 - draw::main::h5d05c67512604564
                               at examples/draw/draw.rs:4
  25:     0x55cd9b7aa22b - std::rt::lang_start::{{closure}}::habdf539bf23fb8b9
                               at /home/julian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  26:     0x55cd9c171cd8 - std::rt::lang_start_internal::{{closure}}::h6d21eebfa4beaf70
                               at src/libstd/rt.rs:52
  27:     0x55cd9c171cd8 - std::panicking::try::do_call::h560a27b87db38b9c
                               at src/libstd/panicking.rs:297
  28:     0x55cd9c171cd8 - std::panicking::try::h453d4afd696011f9
                               at src/libstd/panicking.rs:274
  29:     0x55cd9c171cd8 - std::panic::catch_unwind::h211d02671f23030f
                               at src/libstd/panic.rs:394
  30:     0x55cd9c171cd8 - std::rt::lang_start_internal::h464df2bbf46c7e7c
                               at src/libstd/rt.rs:51
  31:     0x55cd9b7aa207 - std::rt::lang_start::ha15676480cd6900b
                               at /home/julian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  32:     0x55cd9b7acada - main
  33:     0x7f12a07650b3 - __libc_start_main
  34:     0x55cd9b79219e - _start
  35:                0x0 - <unknown>
```

**Comment by conorbros** [comment]
I am also having this issue. Did you find a fix?

**Comment by ambiso** [comment]
I have found a (non-ideal) workaround:

```diff
diff --git a/nannou/src/app.rs b/nannou/src/app.rs
index da5b633..3066602 100644
--- a/nannou/src/app.rs
+++ b/nannou/src/app.rs
@@ -1066,9 +1066,13 @@ fn run_loop<M, E>(
                 };
 
                 if let Some(model) = model.as_ref() {
-                    let swap_chain_output = swap_chain
-                        .get_next_texture()
-                        .expect("failed to acquire next swapchain texture");
+                    let swap_chain_output = loop {
+                        let swap_chain_output = swap_chain.get_next_texture();
+                        match swap_chain_output {
+                            Ok(s) => break s,
+                            Err(_) => {},
+                        }
+                    };
                     let swap_chain_texture = &swap_chain_output.view;
 
                     // Borrow the window now that we don't need it mutably until setting the render
```

I believe the error is rooted in using too many swap chain textures at once.
Maybe one should use an existing one, or wait until one becomes available.
The workaround does the latter, but very inefficiently.

The method `get_next_texture` seems to have been/ renamed to [`get_current_frame` in wgpu-0.6](https://docs.rs/wgpu/0.6.2/wgpu/struct.SwapChain.html#method.get_current_frame).

This method calls [`swap_chain_get_current_texture_view`](https://github.com/gfx-rs/wgpu/blob/7e3965bb5a6f8da0692237fb1fd63fe03434b405/wgpu-core/src/swap_chain.rs#L135) (via wgpu-native), which in turn calls [`acquire_image`](https://github.com/gfx-rs/wgpu/blob/7e3965bb5a6f8da0692237fb1fd63fe03434b405/wgpu-core/src/swap_chain.rs#L158) with a timeout of 1 second. For the vulkan backend (gfx-backend-vulkan) this results in a call to [vkAcquireNextImageKHR](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkAcquireNextImageKHR.html).

I've timed the call, and it instantly returns after less than a millisecond - therefore it's highly unlikely that the timeout is actually reached.

**Comment by ambiso** [comment]
I could not really figure out something useful from the documentation of `vkAcquireNextImageKHR`.

Since I encounter this error on an AMD APU graphics chip, I have analyzed the AMDVLK driver a bit.
Here are some relevant parts:

https://github.com/GPUOpen-Drivers/xgl/blob/2f28502c5e6e4f059a014ab30001a191328f7a5d/icd/api/vk_swapchain.cpp#L1742

https://github.com/GPUOpen-Drivers/xgl/blob/2f28502c5e6e4f059a014ab30001a191328f7a5d/icd/api/vk_swapchain.cpp#L663

https://github.com/GPUOpen-Drivers/pal/blob/85f858393412732198fcd3dc0a0efb8653f5d3f9/src/core/swapChain.cpp#L154

I haven't looked into it too deeply yet, but there's two ways to obtain a timeout error:
https://github.com/GPUOpen-Drivers/xgl/blob/2f28502c5e6e4f059a014ab30001a191328f7a5d/icd/api/include/vk_conv.h#L2454-L2457

**Comment by ambiso** [comment]
If the swapchain is used in the "mailbox" mode, AMDVLK does something similar to what I'm doing right now:

https://github.com/GPUOpen-Drivers/pal/blob/85f858393412732198fcd3dc0a0efb8653f5d3f9/src/core/os/amdgpu/amdgpuSwapChain.cpp#L272-L284

I still have not identified a path that could cause an immediate return from this function.

**Comment by ambiso** [comment]
Here is a workaround that yields the thread upon a timeout error, and handles other errors as before:

```diff
diff --git a/nannou/src/app.rs b/nannou/src/app.rs
index 2812952..52d2107 100644
--- a/nannou/src/app.rs
+++ b/nannou/src/app.rs
@@ -1105,9 +1105,18 @@ fn run_loop<M, E>(
                 };
 
                 if let Some(model) = model.as_ref() {
-                    let swap_chain_output = swap_chain
-                        .get_current_frame()
-                        .expect("failed to acquire next swapchain texture");
+                    let swap_chain_output = loop {
+                        let result = swap_chain.get_current_frame();
+                        match result {
+                            Err(wgpu_upstream::SwapChainError::Timeout) => {
+                                std::thread::yield_now();
+                            },
+                            Err(_) => {
+                                result.expect("failed to acquire next swapchain texture");
+                            },
+                            Ok(s) => break s,
+                        }
+                    };
                     let swap_chain_texture = &swap_chain_output.output.view;
 
                     // Borrow the window now that we don't need it mutably until setting the render
```

**Comment by hashedone** [comment]
I have same problem here, but it occurs only on my nVidia GPU - everything works on Intel GPU. I checked drivers and seems up to date. I cannot run any wgppu examples because of error while rendering:
```
[4.481458 ERROR]()(no module): X11 error: XError {
    description: "BadDrawable (invalid Pixmap or Window parameter)",
    error_code: 9,
    request_code: 149,
    minor_code: 4,
}
```

The @ambiso solution doesn't help - it just blocks forever on he loop..

**Comment by mitchmindtree** [comment]
We have a suspicion that this is solved by #722 - if anyone is still running into this just let us know and we can re-open this.

**Comment by BenBergman** [comment]
This solved the crashes I was seeing in my app.
